Motivation chapter (#7)
FLAGlab/bounded-const-generics-rust: document/chapters/motivation.tex

% !TEX root = ../main.tex
\chapter{Motivation}

> Rust can be seen as a language trying to reduce the number of trade-offs for the programmer, for example: Fast code should not be unsafe and abstractions should not have any performance costs. Rust accomplishes this by having a compiler being able to reason more deeply about the code than usual, it can reason about the lifetime of each variable to avoid dangling pointers, it can reason about mutability to avoid data races. 

Reviewer: ```suggestion
One of the main objectives of Rust is to reduce the number of trade-offs for the programmer. For example, fast code should not be unsafe, and abstractions should not have significant performance costs. Rust accomplishes this by having a compiler that is able able to reason more deeply about the code than. Rust can reason about the lifetime of each variable to avoid dangling pointers, or about mutability to avoid data races. 
```

> However, some improvements could be done in the interaction of the language with constant values. Not necessarily from a performance perspective, but also from an ergonomic perspective. This chapter will explain some aspects of the compilation process that could be improved. Chapter \ref{chapter:proposed_solution} will address the solution for these problems.

Reviewer: However, some improvements could be done in the interaction of the language with constant values. Not necessarily from a performance perspective, but also from an ergonomic perspective. This chapter explains some aspects of the compilation ```and reasoning?``` process```es``` that could be improved. Chapter \ref{chapter:proposed_solution} will address the solution for the problems posit hereinafter.

\label{chapter:motivation}
\section{Trait implementations for arrays}

> In Rust, the programmer has the capability to store data in the heap and the stack. Stack allocated values must have an static size (it must be known at compilation). On the other hand, heap allocated values can have dynamic size, but they can only be accessed using references. The differences between vectors and arrays are a perfect example of this trade-off: Vectors, being heap allocated, can grow or reduce in size during execution, but having nested vectors causes performance issues, because access must be done using nested references. Arrays, being stack allocated, have no performance issues when nesting them, but its size must be fixed during execution.

Reviewer: ```suggestion
In Rust, the programmer has the capability to store data in the heap and the stack. On the one hand, stack allocated values must have an static size (it must be known at compilation). On the other hand, heap allocated values can have dynamic size, but they can only be accessed using references. The differences between vectors and arrays are a perfect example of this trade-off: Vectors, being heap allocated, can grow or reduce in size during execution, but having nested vectors causes performance issues, because access must be done using nested references. Arrays, being stack allocated, have no performance issues when nesting them, but their size must be fixed during execution.
```

> This limitations create some ergonomic problems with arrays: Two arrays with different sizes have different types, even if both store values of the same type. Thus, trait implementations for array types must be done manually for each possible size. Because of this, the standard library only implements traits for array types up to a size of 32, even though arrays are a primitive type of the language. In certain cases vectors are used instead of arrays. Making the code unsuitable for applications running on embedded devices (which may not allow heap allocation) or applications needing multidimensional vectors.

Reviewer: ```suggestion
These limitations create some ergonomic problems with arrays: Two arrays with different sizes have different types, even if both store values of the same type. Thus, trait implementations for array types must be done manually for each possible size. Because of this, the standard library only implements traits for array types up to a size of 32, even though arrays are a primitive type of the language. In certain cases vectors are used instead of arrays. Making the code unsuitable for applications running on embedded devices (which may not allow heap allocation), or applications needing multidimensional vectors.
```

> An example of this limitation can be found on \ref{lst:trait_array}, where the implementation of the trait \texttt{Volatile} cannot be easily extended to other array sizes, even when such implementation does not use the size of the array. 

Reviewer: ```suggestion
An example of this limitation is shown in \ref{lst:trait_array}, where the implementation of the trait \texttt{Volatile} cannot be easily extended to other array sizes, even when such implementation does not use the size of the array. 
```

\begin{listing}
	\begin{minted}{rust} 
    const N: usize = 3;

    trait Volatile {
        fn explode(&self);
    }

    impl<T> Volatile for [T; N] {
        fn explode(&self) {
            for _ in self {
                println!("Boom!")
            }
        }
    }

    fn main() {
        [0i32, 1, 2].explode(); 
        [0i32, 1].explode(); // stderr: no method named `explode` 
                             // found for type `[i32; 2]` in the
                             // current scope
    }
	\end{minted}
    \caption{Even though \texttt{Volatile} is implemented for \texttt{[T; 3]}, it is not for \texttt{[T;2]}.}
  \label{lst:trait_array}
\end{listing}

\section{Static control flow and optimizations}
Rust's compiler is capable of doing static control flow to improve performance execution. If a control flow statement can be resolved during compilation, the compiler will remove the unused expressions of the statement. For example, if the condition of an if-else statement can be resolved during compilation, the compiler will replace the statement with one just including the matching arm of the statement.

> An small illustration of this can be seen on \ref{lst:static_control_flow}, where the array size \texttt{N} is a constant value known during compilation. Given that the expression  \texttt{N > 0} can be evaluated during compilation, the compiler will optimize the function, resulting in code similar to \ref{lst:optimized}. However, doing such optimizations increases compilation time. 

Reviewer: ```suggestion
An small illustration of this can be seen in Listing \ref{lst:static_control_flow}, where the array size \texttt{N} is a constant value known during compilation. Given that the expression  \texttt{N > 0} can be evaluated during compilation, the compiler will optimize the function, resulting in code similar to the code shown in Listing \ref{lst:optimized}. However, doing such optimizations increases compilation time. 
```

> This can be seen as a trade-off between compilation time or code generality. If the programmer chooses \ref{lst:static_control_flow}, the code will work even if \texttt{N} needs to be changed afterwards, if the programmer chooses \ref{lst:optimized}, the code will compile faster. But the programmer can not achieve both at the same time.

Reviewer: ```suggestion
This can be seen as a trade-off between compilation time and code generality. If the programmer chooses to use Listing \ref{lst:static_control_flow}, the code will work even if \texttt{N} needs to be changed afterwards, if the programmer chooses the code in Listing \ref{lst:optimized}, the code will compile faster. But the programmer can not achieve both at the same time.
```
